Move common pitfall documentation to Airflow docs (#15183)
mistercrunch/Airflow: docs/apache-airflow/faq.rst

        )

This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with the execution date 2020-01-01 00:00:00.

.. code-block:: python

        bo = BashOperator(
            task_id='my_task',
            bash_command="echo day={{ ds }}",
            dag=dag
        )

By using the ds macros directly in the template_field, the rendered value results in "day=2020-01-01".

Why ``next_ds`` or ``prev_ds`` might not contain expected values?
------------------------------------------------------------------

- When scheduling DAG, the ``next_ds`` ``next_ds_nodash`` ``prev_ds`` ``prev_ds_nodash`` are calculated using
  ``execution_date`` and ``schedule_interval``. If you set ``schedule_interval`` as ``None`` or ``@once``,
  the ``next_ds``, ``next_ds_nodash``, ``prev_ds``, ``prev_ds_nodash`` values will be set to ``None``.
- When manually triggering DAG, the schedule will be ignored, and ``prev_ds == next_ds == ds``

Task execution interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Why does a sensor never complete?
---------------------------------

When waiting on another task's result, it is recommended to raise the priority of the tasks it will be waiting for.

> Otherwise, scheduling can get deadlocked, with the waiting task taking up all available slots, leaving no slots for the

Reviewer: What do you mean by raise priority of the tasks?

Author: Ah! raising the priority_weight. I will include that to be more specific.

Reviewer: I'm not that comfortable with that recommendation.
priority weight should be used because you want to provide a task priority not because you want to "help" the scheduler to avoid deadlock.
Also the paragraph doesn't consider the mode of the sensor. If you are running in `mode='poke'` the recommendation isn't relevant.

Author: I will remove this faq for now.

execution of the tasks they are waiting for. It is also a good idea for the waiting task to have a later scheduler
interval than the task it is waiting on.

``ExternalTaskSensor`` is an example of a task sensing on the state of a task in another DAG that could potentially run
into this deadlock issue.

What does TemplateNotFound mean?
---------------------------------

> TemplateNotFound errors are usually due to misalignment with user expectations when passing special values to Operator
> that trigger Jinja templating. A common occurrence is with :ref:`BashOperators<howto/operator:BashOperator>`.

Reviewer: ```suggestion
TemplateNotFound errors are usually due to misalignment with user expectations when passing path to operator
that trigger Jinja templating. A common occurrence is with :ref:`BashOperators<howto/operator:BashOperator>`.
```

Author: Good suggestion! I will put that in.

        other_dag_id = f'bar_{i}'
        globals()[other_dag_id] = create_dag(other_dag_id)

Even though Airflow supports multiple DAG definition per python file, dynamically generated or else, it is not
recommended as Airflow would like better isolation between DAGs from a fault and deployment perspective and multiple

> DAGs in the same file goes against that.

Reviewer: Can you clarify a bit about this? This feels a bit harsh I read this paragraph several times I'm not sure I fully understand it. This paragraph is under the question: "How to create DAGs dynamically?" Are you recommending against dynamic DAGs? (as their base code is in one file)

Author: The line came verbatim from the old [Common Pitfall](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=62694614) page.

> We do support more than one DAG definition per python file, but it is not recommended as we would like better isolation between DAGs from a fault and deployment perspective and multiple DAGs in the same file goes against that. For now, make sure that the dag object is in the global namespace : you can use the globals dict as in globals()[dag_id] = DAG(...)

I think the original author wanted to affirm that Airflow does allow a user to define multiple DAGs in a file whether they are statically defining those DAGs or in a loop of some sorts and using the globals symbol table to store the definitions. However, having multiple DAG definition goes against Airflow's mindset/ideal of isolation between DAGs from a fault and deployment perspective.

To sum it up, yes, you can create dynamic DAGs but Airflow does not recommend it. 

Unless we do recommend it (?)

Author: Or should that snippet be placed elsewhere in the documentation?

Author: @eladkal

Reviewer: >To sum it up, yes, you can create dynamic DAGs but Airflow does not recommend it.

:thinking:
@vikramkoka @kaxil @ashb @potiuk Can you share your point of view on this ?

that trigger Jinja templating. A common occurrence is with :ref:`BashOperators<howto/operator:BashOperator>`.

Another commonly missed fact is that the files are resolved relative to where the pipeline file lives. You can add
other directories to the ``template_searchpath`` of the DAG object to allow for other non-relative location.

How to trigger tasks based on another task's failure?
-----------------------------------------------------

For tasks that are related through dependency, you can set the ``trigger_rule`` to ``TriggerRule.ALL_FAILED`` if the
task execution depends on the failure of ALL its upstream tasks or ``TriggerRule.ONE_FAILED`` for just one of the
upstream task.

.. code-block:: python

    from airflow.models import DAG
    from airflow.operators.python import PythonOperator
    from airflow.utils.trigger_rule import TriggerRule

    parent = PythonOperator(
        task_id='a',
        python_callable=lambda x: 1,
        dag=dag
    )

    child = PythonOperator(
        task_id='b',
        python_callable=lambda: 1,
        trigger_rule=TriggerRule.ALL_FAILED,
        dag=dag
    )

>     parent >> child

Reviewer: lets use Airflow 2 syntax with dag & task decorators.

Author: Sounds good! Will convert to task decorator.

            }
        )

        bo = BashOperator(
            task_id='my_task',
            bash_command="echo {{ my_custom_macro }}",
            dag=dag
        )

This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with the execution date 2020-01-01 00:00:00.

.. code-block:: python

        bo = BashOperator(
            task_id='my_task',
            bash_command="echo day={{ ds }}",
            dag=dag
        )

By using the ds macros directly in the template_field, the rendered value results in "day=2020-01-01".

Why ``next_ds`` or ``prev_ds`` might not contain expected values?
------------------------------------------------------------------

- When scheduling DAG, the ``next_ds`` ``next_ds_nodash`` ``prev_ds`` ``prev_ds_nodash`` are calculated using
  ``execution_date`` and ``schedule_interval``. If you set ``schedule_interval`` as ``None`` or ``@once``,
  the ``next_ds``, ``next_ds_nodash``, ``prev_ds``, ``prev_ds_nodash`` values will be set to ``None``.
- When manually triggering DAG, the schedule will be ignored, and ``prev_ds == next_ds == ds``

Task execution interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> What does TemplateNotFound mean?
> ---------------------------------

Reviewer 2: ```suggestion
What does ``TemplateNotFound`` mean?
-------------------------------------
```

> TemplateNotFound errors are usually due to misalignment with user expectations when passing path to operator

Reviewer 2: ```suggestion
``TemplateNotFound`` errors are usually due to misalignment with user expectations when passing path to operator
```

other directories to the ``template_searchpath`` of the DAG object to allow for other non-relative location.

How to trigger tasks based on another task's failure?
-----------------------------------------------------

For tasks that are related through dependency, you can set the ``trigger_rule`` to ``TriggerRule.ALL_FAILED`` if the
task execution depends on the failure of ALL its upstream tasks or ``TriggerRule.ONE_FAILED`` for just one of the
upstream task.

.. code-block:: python

    from airflow.decorators import dag, task
    from airflow.exceptions import AirflowException
    from airflow.utils.trigger_rule import TriggerRule

    from datetime import datetime

    @task
    def a_func():
        raise AirflowException

    @task(
        trigger_rule=TriggerRule.ALL_FAILED,
    )
    def b_func():
        pass

    @dag(
        schedule_interval='@once',
        start_date=datetime(2021, 1, 1)
    )
    def my_dag():
        a = a_func()
        b = b_func()

        a >> b

    dag = my_dag()

See :ref:`concepts:trigger-rules` for more information.

If the tasks are not related by dependency, you will need to :ref:`build a custom Operator<custom_operator>`.

> 
> 

Reviewer 2: ```suggestion
```

> Airflow CLI
> ^^^^^^^^^^^
> 
> What are all the ``airflow tasks run`` commands in the process list?
> ---------------------------------------------------------------------
> 
> There are many layers of ``airflow tasks run`` commands, meaning it can call itself.
> 
> - Basic ``airflow tasks run``: fires up an executor, and tell it to run an
>   ``airflow tasks run --local`` command. If using Celery, this means it puts a
>   command in the queue for it to run remotely on the worker. If using
>   LocalExecutor, that translates into running it in a subprocess pool.
> - Local ``airflow tasks run --local``: starts an ``airflow tasks run --raw``
>   command (described below) as a subprocess and is in charge of
>   emitting heartbeats, listening for external kill signals
>   and ensures some cleanup takes place if the subprocess fails.
> - Raw ``airflow tasks run --raw`` runs the actual operator's execute method and
>   performs the actual work.
> 

Reviewer 2: Let's move this to separate PR plz

```suggestion
```

Airflow UI
^^^^^^^^^^

How do I stop the sync perms happening multiple times per webserver?
--------------------------------------------------------------------

Set the value of ``update_fab_perms`` configuration in ``airflow.cfg`` to ``False``.

How to reduce the airflow UI page load time?
------------------------------------------------

If your dag takes long time to load, you could reduce the value of ``default_dag_run_display_number`` configuration
in ``airflow.cfg`` to a smaller value. This configurable controls the number of dag run to show in UI with default
value ``25``.

Why did the pause dag toggle turn red?
--------------------------------------

If pausing or unpausing a dag fails for any reason, the dag toggle will
revert to its previous state and turn red. If you observe this behavior,
try pausing the dag again, or check the console or server logs if the
issue recurs.

MySQL and MySQL variant Databases
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

What does "MySQL Server has gone away" mean?
--------------------------------------------

> You may occasionally experience OperationalError with the message "MySQL Server has gone away". This is due to the

Reviewer 2: ```suggestion
You may occasionally experience ``OperationalError`` with the message "MySQL Server has gone away". This is due to the
```

connection pool keeping connections open too long and you are given an old connection that has expired. To ensure a
valid connection, you can set :ref:`config:core__sql_alchemy_pool_recycle` to ensure connections are invalidated after
that many seconds and new ones are created.

Does Airflow support extended ASCII or unicode characters?
----------------------------------------------------------

If you intend to use extended ASCII or Unicode characters in Airflow, you have to provide a proper connection string to
the MySQL database since they define charset explicitly.

.. code-block:: text

    sql_alchemy_conn = mysql://airflow@localhost:3306/airflow?charset=utf8

> You will experience UnicodeDecodeError thrown by WTForms templating and other Airflow modules like below.

Reviewer 2: ```suggestion
You will experience ``UnicodeDecodeError`` thrown by ``WTForms`` templating and other Airflow modules like below.
```

.. code-block:: text

   'ascii' codec can't decode byte 0xae in position 506: ordinal not in range(128)

> How to fix Exception: Global variable explicit_defaults_for_timestamp needs to be on (1)?
> -----------------------------------------------------------------------------------------

Reviewer 2: ```suggestion
How to fix Exception: Global variable ``explicit_defaults_for_timestamp`` needs to be on (1)?
---------------------------------------------------------------------------------------------
```
